### PR TITLE
Refactor to fix all type problems

### LINKED ISSUE #4496: Fix all types problems
During migration from Flow to TypeScript (https://github.com/stylelint/stylelint/issues/4245) we marked many places with `@ts-ignore` comments. We need to remove all these comments and fix types problems, which were suppressed by these comments.

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Closes #4496

> Is there anything in the PR that needs further explanation?

This change:

- fixes all `// @ts-nocheck` comments
- fixes all `// TODO TYPES` comments (some are replaced with `@ts-expect-error`)

### CODE REVIEW

File: lib/formatters/index.js [hunk 1]
```diff
@@ -1,15 +1,17 @@
 'use strict';
 
-const importLazy = require('import-lazy');
+const _importLazy = require('import-lazy');
+
+const importLazy = _importLazy(require);
```
Comment (ybiquitous) [inline lib/formatters/index.js:5]: [note] The same way to use `import-lazy` as `lib/rules/index.js`

File: lib/rules/time-min-milliseconds/index.js [hunk 2]
```diff
@@ -101,34 +101,54 @@ function rule(minimum, options) {
...
+		 * @param {string} time
+		 * @returns {boolean}
+		 */
 		function isAcceptableTime(time) {
 			const parsedTime = valueParser.unit(time);
 
 			if (!parsedTime) return true;
 
-			if (parsedTime.number <= 0) {
+			const numTime = Number(parsedTime.number);
```
Comment (ybiquitous) [inline lib/rules/time-min-milliseconds/index.js:128]: [note] Avoiding implicit conversion from string to number (`parsedTime.number` is a string).

File: lib/rules/value-list-comma-newline-after/index.js [hunk 3]
```diff
@@ -60,7 +60,7 @@ function rule(expectation, options, context) {
 				  }
 				: null,
 			determineIndex: (declString, match) => {
-				const nextChars = declString.substr(match.endIndex, declString.length - match.endIndex);
+				const nextChars = declString.substring(match.endIndex, declString.length);
```
Comment (ybiquitous) [inline lib/rules/value-list-comma-newline-after/index.js:63]: [note] [`substr()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) is deprecated.

File: lib/utils/validateTypes.js [hunk 4]
```diff
@@ -36,9 +36,43 @@ function isString(value) {
...
+
 module.exports = {
 	isBoolean,
 	isNumber,
 	isRegExp,
 	isString,
+
+	assert,
+	assertNumber,
+	assertString,
```
Comment (ybiquitous) [inline lib/utils/validateTypes.js:77]: [ask] Is here valid as the location of Assertion Functions?

File: lib/utils/validateTypes.js [hunk 5]
```diff
@@ -36,9 +36,43 @@ function isString(value) {
...
 }
 
+/**
+ * Assert that the value is truthy.
+ * @param {unknown} value
+ * @returns {asserts value}
+ */
+function assert(value) {
+	// eslint-disable-next-line no-console
+	console.assert(value);
```
Comment (ybiquitous) [inline lib/utils/validateTypes.js:46]: [note] [`console.assert()`](https://developer.mozilla.org/en-US/docs/Web/API/console/assert) is available both on Node.js and modern browsers.

**Review by jeddy3 — APPROVED** [review]
Amazing work.

Thank you for the huge effort you've put into fixing all the type problems!

LGTM.

**Comment by ybiquitous** [comment]
@jeddy3 Thank you too for your review. 😊

**Comment by Airkro** [comment]
This PR break `stylelint` bundling.

`rollup` or `webpack` can't build `stylelint` correctly because `importLazy`.

Any solution?

**Comment by jeddy3** [comment]
@Airkro Please open a [new bug report](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REPORT_A_BUG.yml) so that we can gather more details, track and discuss the issue.

**Pull request merged**